cloudevents/spec.md

```diff
@@ -436,15 +436,27 @@ on the definition of OPTIONAL.
 - Constraints:
   - OPTIONAL
   - If present, MUST be a non-empty string
-- Example:
+- Examples:
   - A subscriber might register interest for when new blobs are created inside a
     blob-storage container. In this case, the event `source` identifies the
     subscription scope (storage container), the `type` identifies the "blob
     created" event, and the `id` uniquely identifies the event instance to
     distinguish separate occurrences of a same-named blob having been created;
     the name of the newly created blob is carried in `subject`:
-    - `source`: `https://example.com/storage/tenant/container`
-    - `subject`: `mynewfile.jpg`
+      - `source`: `https://example.com/storage/tenant/container`
+      - `subject`: `mynewfile.jpg`
+  - A subscriber might register interest for when new updates are made to a client
+    in an eCommerce system. In this case, the event `source` identifies the
+    subscription scope (CRM part of an eCommerce system), the `type` identifies
+    the "client updated" event, and the `id` uniquely identifies the event
+    instance to distinguish separate occurrences of a same client being
+    updated multiple times; the affiliate partner id and client id of the updated
+    client are carried in `subject`. Here, it is a composite string
+    (`affiliate partner id:client id`) that is made up of one or more identifying
+    pieces of data separated by some symbol (like, colon), assuming that each
+    affiliate partner has its own client base:
```

Comment:
I'd prefer if it said something like (on lines 453-457):
- - -
the `subject` uniquely identifies the client within the scope of the `source` by including a "partner id" and "client id" (which is unique within the scope of the "partner id") separated by a colon.
- - -

This makes it clear that the value can be a composite value, but also makes it clear that each sub-part is there to help uniquely identify the subject and not just other interesting info.

TBH I'm still not sure all of this is needed since the same concerns could be said about any other attribute that identifies something (like "source" or even "id" itself) - those can be composites too.

Reply:
> I'd prefer if it said something like (on lines 453-457):

Done.

> TBH I'm still not sure all of this is needed since the same concerns could be said about any other attribute that identifies something (like "source" or even "id" itself) - those can be composites too.

@duglin Don't forget that I've also fixed the word _describes_ into _identifies_ at line 422, an inconsistency in the specification that has generated a lots of confusion. Consequently, this extra example would help people understand the true intention behind `subject` and how to cope with a composite content.

The `source` is by its very nature following a composite (hierarchical) structure since it is an URI reference. An `id` is usually opaque and rarely, if ever, parsed, so its content doesn't really matter except for providing uniqueness.

Reply:
yes I do agree with the s/describes/identifies/ change :-)

Reply:
> yes I do agree with the s/describes/identifies/ change :-)

@duglin As the proverb goes: _"Little by little, a little becomes a lot.”_ At the end, I am glad that all these _little_ messages in this PR have revealed a _lot_, that assumptions about `subject` were not fully aligned with facts from the specification. :-)